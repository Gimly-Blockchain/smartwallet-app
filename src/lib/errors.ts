import {routeList} from '../routeList'
import strings from '../locales/strings'
import ErrorCode from './errorCodes'

export class AppError extends Error {
  // private code: ErrorCode
  public origError: any
  navigateTo: routeList

  public constructor(code = ErrorCode.Unknown, origError?: any, navigateTo: routeList = routeList.Home) {
    super(strings[code] || strings[ErrorCode.Unknown])
    // this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]
