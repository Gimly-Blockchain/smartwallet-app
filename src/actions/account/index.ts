import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import { CategorizedClaims, DecoratedClaims } from 'src/reducers/account'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
} from '../../lib/util'
import { cancelReceiving } from '../sso'
import { JolocomLib } from 'jolocom-lib'
import { ThunkAction } from 'src/store'
import { groupBy, map, mergeRight, omit, uniq, zipWith } from 'ramda'
import { compose } from 'redux'
import { CredentialMetadataSummary } from '../../lib/storage/storage'
import { IdentitySummary } from '../sso/types'
import { KeyTypes } from 'jolocom-lib/js/vaultedKeyProvider/types'
import { publicKeyToDID } from 'jolocom-lib/js/utils/crypto'
import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { Not } from 'typeorm'
import { HAS_EXTERNAL_CREDENTIALS } from './actionTypes'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANDLE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const checkIdentityExists: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { keyChainLib, storageLib, encryptionLib, registry } = backendMiddleware
  const encryptedEntropy = await storageLib.get.encryptedSeed()

  if (!encryptedEntropy) {
    const isRegistering = getState().registration.loading.isRegistering

    const routeName = isRegistering
      ? routeList.RegistrationProgress
      : routeList.Landing

    return dispatch(navigationActions.navigate({ routeName }))
  }

  const encryptionPass = await keyChainLib.getPassword()

  const decryptedSeed = encryptionLib.decryptWithPass({
    cipher: encryptedEntropy,
    pass: encryptionPass,
  })

  if (!decryptedSeed) {
    throw new Error('could not decrypt seed')
  }

  // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
  const userVault = JolocomLib.KeyProvider.fromSeed(
    Buffer.from(decryptedSeed, 'hex'),
    encryptionPass,
  )

  const userPubKey = userVault.getPublicKey({
    derivationPath: KeyTypes.jolocomIdentityKey,
    encryptionPass,
  })

  const didDocument = await storageLib.get.didDoc(publicKeyToDID(userPubKey))

  if (didDocument) {
    const identity = Identity.fromDidDocument({ didDocument })

    // TODO Simplify constructor
    backendMiddleware.identityWallet = new IdentityWallet({
      identity,
      vaultedKeyProvider: userVault,
      publicKeyMetadata: {
        derivationPath: KeyTypes.jolocomIdentityKey,
        keyId: identity.publicKeySection[0].id,
      },
      contractsAdapter: registry.contractsAdapter,
      contractsGateway: registry.contractsGateway,
    })
  } else {
    const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes
    const identityWallet = await registry.authenticate(userVault, {
      encryptionPass,
      derivationPath,
    })

    backendMiddleware.identityWallet = identityWallet
    await storageLib.store.didDoc(identityWallet.didDocument)
  }

  const userDid = backendMiddleware.identityWallet.identity.did
  dispatch(setDid(userDid))

  return dispatch(navigationActions.navigate({ routeName: routeList.Home }))
}

export const openClaimDetails = (
  claim: DecoratedClaims,
): ThunkAction => dispatch => {
  dispatch(setSelected(claim))
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { identityWallet, storageLib, keyChainLib } = backendMiddleware

  const did = getState().account.did.did
  const claimsItem = getState().account.claims.selected
  const password = await keyChainLib.getPassword()

  const verifiableCredential = await identityWallet.create.signedCredential(
    {
      metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
      // the library acts directly on the object passed in, so a copy should be made first
      claim: { ...claimsItem.claimData },
      subject: did,
    },
    password,
  )

  if (claimsItem.id) {
    await storageLib.delete.verifiableCredential(claimsItem.id)
  }

  await storageLib.store.verifiableCredential(verifiableCredential)

  await dispatch(setClaimsForDid)

  return dispatch(navigationActions.navigatorResetHome())
}

// TODO Currently only rendering  / adding one
export const saveExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib } = backendMiddleware
  const externalCredentials = getState().account.claims.pendingExternal

  if (!externalCredentials.offer.length) {
    return dispatch(cancelReceiving)
  }

  const cred: SignedCredential = externalCredentials.offer[0].credential

  await storageLib.delete.verifiableCredential(cred.id)
  await storageLib.store.verifiableCredential(cred)

  return dispatch(cancelReceiving)
}

export const toggleLoading = (value: boolean) => ({
  type: 'SET_LOADING',
  value,
})

export const hasExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib, identityWallet } = backendMiddleware
  const externalCredentials = await storageLib.get.verifiableCredential({
    issuer: Not(identityWallet.did),
  })

  return dispatch({
    type: HAS_EXTERNAL_CREDENTIALS,
    value: externalCredentials.length !== 0,
  })
}
export const setClaimsForDid: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib } = backendMiddleware

  const verifiableCredentials: SignedCredential[] = await storageLib.get.verifiableCredential()

  const metadata = await Promise.all(
    verifiableCredentials.map(el => storageLib.get.credentialMetadata(el)),
  )

  const issuers = uniq(verifiableCredentials.map(cred => cred.issuer))

  const issuerMetadata = await Promise.all(
    issuers.map(storageLib.get.publicProfile),
  )

  const claims = prepareClaimsForState(
    verifiableCredentials,
    metadata,
    issuerMetadata,
  ) as CategorizedClaims

  return dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })
}

export const prepareClaimsForState = (
  credentials: SignedCredential[],
  credentialMetadata: Array<CredentialMetadataSummary | {}>,
  issuerMetadata: Array<IdentitySummary | { did: string }>,
) =>
  compose(
    groupBy(getCredentialUiCategory),
    zipWith(mergeRight, credentialMetadata),
    map(addIssuerInfo(issuerMetadata)),
    map(convertToDecoratedClaim),
  )(credentials)

export const addIssuerInfo = (
  issuerProfiles: Array<{ did: string } | IdentitySummary> | [],
) => (claim: DecoratedClaims) => {
  if (!issuerProfiles || !issuerProfiles.length) {
    return claim
  }

  const issuer = issuerProfiles.find(el => el.did === claim.issuer.did)

  return issuer
    ? {
        ...claim,
        issuer,
      }
    : claim
}

/** @TODO Util, make subject mandatory (in lib) */
export const convertToDecoratedClaim = ({
  claim,
  type,
  issuer,
  id,
  expires,
}: SignedCredential): DecoratedClaims => ({
  credentialType: getUiCredentialTypeByType(type),
  issuer: {
    did: issuer,
  },
  claimData: omit(['id'], claim),
  id,
  subject: claim.id || 'Not found',
  expires: expires || undefined,
})
