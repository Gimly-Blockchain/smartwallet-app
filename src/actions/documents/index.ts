import { Action, AnyAction } from 'redux'
import { Dispatch } from 'react-redux'
import { navigationActions } from '..'
import { routeList } from 'src/routeList'
import I18n from 'src/locales/i18n'
import { DecoratedClaims } from 'src/reducers/account'
import { ClaimInterface } from 'cred-types-jolocom-core'

export const SET_EXPIRED_DOC = 'SET_SELECTED_EXPIRED_DOCUMENT'
export const CLEAR_EXPIRED_DOC = 'CLEAR_SELECTED_EXPIRED_DOCUMENT'

export const setSelectedExpiredDocument = (
  document: DecoratedClaims,
): AnyAction => ({
  type: SET_EXPIRED_DOC,
  value: document,
})

export const clearSelectedExpiredDocument = (): Action => ({
  type: CLEAR_EXPIRED_DOC,
})

export const openExpiredDetails = (document: DecoratedClaims) => async (
  dispatch: Dispatch<AnyAction>,
) => {
  dispatch(setSelectedExpiredDocument(document))
  dispatch(
    navigationActions.navigate({
      routeName: routeList.ExpiredDetails,
      params: {
        headerTitle: `${I18n.t('[expired]')} ${
          (document.claimData as ClaimInterface).type
        }`,
      },
    }),
  )
}
