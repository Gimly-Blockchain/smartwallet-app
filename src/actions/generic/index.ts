import { Linking } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { routeList } from 'src/routeList'
import { navigationActions, accountActions } from 'src/actions'
import { ThunkAction } from 'src/store'
import settingKeys from 'src/ui/settings/settingKeys'
import { withLoading, withErrorScreen } from '../modifiers'
import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
} from 'src/reducers/generic'
import { AnyAction } from 'redux'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300
import { AppError, ErrorCode } from '@jolocom/sdk/js/src/lib/errors'
import { AsyncStorage } from 'react-native'
import crypto from 'crypto'
import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()

    // locale setup
    /** locale setup
     * @dev Until German and Dutch terms are polished, only English is used.
     * previous code:
     */
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale

    storedSettings.locale = I18n.locale

    await dispatch(loadSettings(storedSettings))

    const ret = await dispatch(accountActions.checkIdentityExists)

    // FIXME what happens if no identity and this is a deeplink?
    // navigationActions.handleDeepLink throws NoWallet
    // need to improve this UX

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    await Linking.getInitialURL().then(url => {
      if (url) handleDeepLink(url)
    })

    return ret
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  } finally {
    SplashScreen.hide()
  }
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()
  return dispatch(
    navigationActions.navigate({
      routeName: shouldShowTerms ? routeList.TermsOfServiceConsent : route,
      params: { nextRoute: route, onSubmit },
    }),
  )
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: 'LOAD_SETTINGS',
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
