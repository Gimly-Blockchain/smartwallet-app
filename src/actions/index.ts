import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as interactionHandlerActions from 'src/actions/interactionHandler'
import * as paymentActions from 'src/actions/payment/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  interactionHandlerActions,
  paymentActions
}
