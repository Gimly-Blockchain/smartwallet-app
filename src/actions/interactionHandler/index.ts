import { Dispatch, AnyAction } from 'redux'
import { JolocomLib } from 'jolocom-lib'
import { showErrorScreen } from 'src/actions/generic'
import { InteractionType } from 'jolocom-lib/js/interactionTokens/types'
import { accountActions, ssoActions } from 'src/actions'

export const parseJWT = (encodedJwt: string) => {
  return async (dispatch: Dispatch<AnyAction>) => {
    dispatch(accountActions.toggleLoading(true))
    try {
      const returnedDecodedJwt = await JolocomLib.parse.interactionToken.fromJWT(encodedJwt)
      
      if (returnedDecodedJwt.interactionType === InteractionType.CredentialRequest) {
        dispatch(ssoActions.consumeCredentialRequest(returnedDecodedJwt))
      }
      if (returnedDecodedJwt.interactionType === InteractionType.CredentialOffer) {
        dispatch(ssoActions.consumeCredentialOfferRequest(returnedDecodedJwt))
      }
      if (returnedDecodedJwt.interactionType === InteractionType.CredentialsReceive) {
        dispatch(ssoActions.receiveExternalCredential(returnedDecodedJwt))
      }
    } catch (err) {
      console.log('error: ', err)
      dispatch(accountActions.toggleLoading(false))
      dispatch(showErrorScreen(new Error('JWT Token parse failed')))
    }
  }
}
