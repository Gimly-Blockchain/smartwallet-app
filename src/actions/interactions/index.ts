import { ThunkAction } from 'src/store';

export const connectToBleDevice = (device) => {
}

const ADD_BLE_DEVICE = 'BLA'

export const addBleDevice = (device) => ({
  type: ADD_BLE_DEVICE,
  value: { id: device.id, name: device.name }
})

export const initiateBLEScan: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware
) => {
  backendMiddleware.bleLib.startDeviceScan(device => dispatch(addBleDevice(device)))
}
