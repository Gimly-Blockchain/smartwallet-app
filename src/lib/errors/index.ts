import { routeList } from 'src/routeList'
import strings from '../../locales/strings'
import { ErrorCode } from '@jolocom/sdk/js/src/lib/errors/types'
export { ErrorCode }

export interface IAppError {
  origError: any
  navigateTo: routeList
}
export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

export interface ErrorReport extends UserReport {
  error: IAppError | Error | undefined
}
import { reportErrorToSentry, initSentry } from 'src/lib/errors/sentry'

export class AppError extends Error implements IAppError {
  // private code: ErrorCode
  public origError: any
  public navigateTo: routeList

  public constructor(
    code = ErrorCode.Unknown,
    origError?: any,
    navigateTo: routeList = routeList.AppInit,
  ) {
    super(strings[code] || strings[ErrorCode.Unknown])
    // this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]

export function reportError(report: ErrorReport) {
  let extraData
  if (report.error instanceof AppError && report.error.origError) {
    extraData = {
      AppError: report.error.message,
      UserError: report.userError,
      UserDescription: report.userDescription,
      UserContact: report.userContact,
      sendPrivateData: report.sendPrivateData,
    }
    report.error = report.error.origError
  }
  return reportErrorToSentry(report, extraData)
}

export const initErrorReporting = initSentry
