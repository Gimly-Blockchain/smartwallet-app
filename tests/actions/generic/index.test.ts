import {genericActions} from 'src/actions'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import {AppError} from '../../../src/lib/errors'
import ErrorCode from '../../../src/lib/errorCodes'

describe('Generic action creators', () => {
  describe('showErrorScreen', () => {
    it('should navigate to error screen and provide the message', () => {
      const mockStore = configureStore([thunk])({})
      const mockError = new AppError(ErrorCode.Unknown,  new Error('Test Error'))

      const action = genericActions.showErrorScreen(mockError)

      mockStore.dispatch(action)
      expect(mockStore.getActions()).toMatchSnapshot()
    })
  })
})
