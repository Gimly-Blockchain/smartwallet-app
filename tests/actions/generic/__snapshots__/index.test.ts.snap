// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic action creators showErrorScreen should navigate to error screen and provide the message 1`] = `
Array [
  Object {
    "params": Object {
      "error": Object {
        "message": "MOCK BAD ERROR",
        "stack": "MOCK STACK TRACE",
      },
      "returnTo": "Home",
    },
    "routeName": "Exception",
    "type": "Navigation/NAVIGATE",
  },
]
`;
